Index: lightdm/tests/src/libsystem.c
===================================================================
--- lightdm.orig/tests/src/libsystem.c
+++ lightdm/tests/src/libsystem.c
@@ -329,6 +329,9 @@ stat64 (const char *path, struct stat64
     return _stat64 (new_path, buf);
 }
 
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-prototypes"
+
 int
 __xstat (int version, const char *path, struct stat *buf)
 {
@@ -365,6 +368,8 @@ __fxstatat64(int ver, int dirfd, const c
     return ___fxstatat64 (ver, dirfd, new_path, buf, flags);
 }
 
+#pragma GCC diagnostic pop
+
 DIR *
 opendir (const char *name)
 {
