Index: lightdm/src/session-child.c
===================================================================
--- lightdm.orig/src/session-child.c
+++ lightdm/src/session-child.c
@@ -17,6 +17,7 @@
 #include <utmpx.h>
 #include <sys/mman.h>
 #include <pthread.h>
+#include <locale.h>
 
 #if HAVE_LIBAUDIT
 #include <libaudit.h>
@@ -367,6 +368,8 @@ session_child_run (int argc, char **argv
     g_autofree gchar *xdisplay = read_string ();
     g_autoptr(XAuthority) x_authority = read_xauth ();
 
+    setlocale (LC_ALL, "");
+
     /* Setup PAM */
     struct pam_conv conversation = { pam_conv_cb, NULL };
     int result = pam_start (service, username, &conversation, &pam_handle);
